### PR TITLE
Substructure: Atom data and boolean tree + tests

### PR DESCRIPTION
**Pull request by scal444**
First PR for substructure search. Includes definitions for bit-packed atom data and comparison functions, and an implementation of a boolean tree walk. 

The boolean tree reference recursive bits that are not yet implemented, but can be tested with dummy data.

### CODE REVIEW

File: src/substruct/atom_data_packed.h [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+  static constexpr int kDegreeBits    = 6;     ///< Number of bits for degree field (max 63)
+  static constexpr int kDegreeMask    = 0x3F;  ///< Mask for 6-bit degree (bits 0-5)
+  static constexpr int kIsAromaticBit = 6;     ///< Bit position within degree byte for isAromatic
+  static constexpr int kIsInRingBit   = 7;     ///< Bit position within degree byte for isInRing
+  /// @}
+
+  /// @name Compacted 4-bit field masks and bit positions
+  /// @{
+  static constexpr int k4BitMask              = 0x0F;  ///< Mask for 4-bit fields (max value 15)
+  static constexpr int kNumRingsBits          = 0;     ///< Low 4 bits of byte 1
```
Comment (evasnow1992) [inline src/substruct/atom_data_packed.h:106]: I notice for several atom properties (numRings, ringBondCount, numImplicitHs, numHeteroatomNeighbors), we're using 4-bit fields with a max value of 15. Should we add bounds checking or saturation in the setters? Currently, values > 15 are silently truncated, which could cause unexpected behavior.

**Review by evasnow1992 — APPROVED** [review]
Thank you for adding packed atom data and boolean tree operations. Only one minor comment. Also may considering adding some edge case tests like for empty trees, and/or tests for comparison operations like GreaterThan. Not required though

File: src/substruct/atom_data_packed.h (hunk 1 above)
Comment (scal444) [reply src/substruct/atom_data_packed.h:106]: I've got those checks in the code that uses this in one of the next PRs
